tcmur_cmd_handler.c

```diff
@@ -294,6 +298,618 @@ static int handle_write_verify(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
+		return tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+					   ASC_INVALID_FIELD_IN_PARAMETER_LIST,
+					   NULL);
+	}
+
+	/*
+	 * Source designator matches the local device
+	 */
+	if (index == xcopy->stdi) {
+		memcpy(&xcopy->src_tid_wwn[0], &tgt_desc[8],
+		       XCOPY_NAA_IEEE_REGEX_LEN);
+
+		if (!memcmp(wwn, xcopy->src_tid_wwn, XCOPY_NAA_IEEE_REGEX_LEN))
+			xcopy->src_dev = udev;
+	}
+
+	/*
+	 * Destination designator matches the local device.
+	 */
+	if (index == xcopy->dtdi) {
+		memcpy(xcopy->dst_tid_wwn, &tgt_desc[8],
+		       XCOPY_NAA_IEEE_REGEX_LEN);
+
+		if (!memcmp(wwn, xcopy->dst_tid_wwn, XCOPY_NAA_IEEE_REGEX_LEN))
+			xcopy->dst_dev = udev;
+	}
+
+	return SAM_STAT_GOOD;
+}
+
+static int xcopy_parse_target_descs(struct tcmu_device *udev,
+				    struct xcopy *xcopy,
+				    uint8_t *tgt_desc,
+				    uint16_t tdll,
+				    uint8_t *sense)
+{
+	int i, ret;
+
+	if (tdll > RCR_OP_MAX_TARGET_DESC_COUNT * XCOPY_TARGET_DESC_LEN) {
+		tcmu_dev_err(udev, "Only %u target descriptor(s) supported, but there are %u\n",
+			     RCR_OP_MAX_TARGET_DESC_COUNT, tdll / XCOPY_TARGET_DESC_LEN);
+		return tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+					   ASC_INVALID_FIELD_IN_PARAMETER_LIST,
+					   NULL);
+	}
+
+	for (i = 0; i < RCR_OP_MAX_TARGET_DESC_COUNT; i++) {
+		/*
+		 * Only Identification Descriptor Target Descriptor support
+		 * for now.
+		 */
+		if (tgt_desc[0] == XCOPY_TARGET_DESC_TYPE_CODE_ID) {
+			ret = xcopy_parse_target_id(udev, xcopy, tgt_desc, i, sense);
+			if (ret != SAM_STAT_GOOD)
+				return ret;
+
+			tgt_desc += XCOPY_TARGET_DESC_LEN;
+		} else {
+			tcmu_dev_err(udev, "Unsupport target descriptor type code 0x%x\n",
+				     tgt_desc[0]);
+			return tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+						   ASC_UNSUPPORTED_TARGET_DESC_TYPE_CODE,
+						   NULL);
+		}
+	}
+
+	if (xcopy->src_dev)
+		ret = xcopy_locate_udev(udev->ctx, xcopy->dst_tid_wwn,
+					&xcopy->dst_dev);
+	else if (xcopy->dst_dev)
+		ret = xcopy_locate_udev(udev->ctx, xcopy->src_tid_wwn,
+					&xcopy->src_dev);
+	else
+		ret = SAM_STAT_CHECK_CONDITION;
+
+	if (ret != SAM_STAT_GOOD) {
+		tcmu_err("Target device not found, the index are %hu and %hu\n",
+			 xcopy->stdi, xcopy->dtdi);
+		return tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+					   ASC_UNSUPPORTED_TARGET_DESC_TYPE_CODE,
+					   NULL);
+	}
+
+	tcmu_dev_dbg(xcopy->src_dev, "Source device NAA IEEE WWN: 0x%16phN\n",
+		     xcopy->src_tid_wwn);
+	tcmu_dev_dbg(xcopy->dst_dev, "Destination device NAA IEEE WWN: 0x%16phN\n",
+		     xcopy->dst_tid_wwn);
+
+	return SAM_STAT_GOOD;
+}
+
+static int xcopy_parse_parameter_list(struct tcmu_device *dev,
+				      struct tcmulib_cmd *cmd,
+				      struct xcopy *xcopy)
+{
+	uint8_t *cdb = cmd->cdb;
+	size_t data_length = tcmu_get_xfer_length(cdb);
+	struct iovec *iovec = cmd->iovec;
+	size_t iov_cnt = cmd->iov_cnt;
+	uint8_t *sense = cmd->sense_buf;
+	uint32_t inline_dl;
+	uint8_t *seg_desc, *tgt_desc, *par;
+	uint16_t sdll, tdll;
+	int ret;
+
```

Comment:
I guess xcopy has a similar problem as PGRs where because we do not know the path a command came in on, a backend device can only be exported through one enabled port per node until (ALUA has a similar issue, but we only need to fix the inquiry code). I am ok with this restriction for now. It probably will not be a big deal, but we cannot support the LIST IDENTIFIER field because that is per I_T Nexus right?

I think we need to make sure LIST ID USAGE == 11b and we need to check that if LIST IDENTIFIER is not zero return an INVAlID FIELD IN PARAM LIST check condition.


xcopy has a similar issue as PGRs right? We do not know the I_T Nexus (iSCSI session in your case) that commands came in on, so the backend device can only be exported through one enabled port per target (or per node for HA). If it is exported through multiple ports on one target/node then the list identifier could clash but be ok since the ID is per I_T Nexus.

So I think we have to either do some craziness like IBM is doing with PGRs where the kernel does the SCSI/transport checks and then pushes actual IO processing to runner, or just make the above restriction for now which seems ok.

If we are saying it will be a limited with one enabled port per node per backend, then